Integrating StyleSystem Classes for Host Container tags in SPA Editor - Angular Impl

friendlymahi opened the pull request:
## Description

Changes done to EditableComponent to support StyleSystem based on appliedCssClassNames property from Sling Model exporters. Please refer to related issue for more details.

## Related Issue
Here is the related issue requesting this feature. #60 


## Motivation and Context

These changes extend the style system capability/feature to SPA Editor implementation for Angular apps.

## How Has This Been Tested?

Changes done locally and deployed to AEM instance via a brand new project to confirm that style-system works as expected. Additionally unit tests are also included to validate this behavior.

## Screenshots (if appropriate):
![Screen Shot 2021-08-16 at 8 17 08 PM](https://user-images.githubusercontent.com/34920377/129649915-25550e3c-8db3-44cc-a66b-c397e8490120.png)
![Screen Shot 2021-08-16 at 8 28 48 PM](https://user-images.githubusercontent.com/34920377/129649918-a2524f52-d244-417d-9a55-6752678a58ab.png)


## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

## Additional Notes

If this PR is approved/merged, the page below requires updates against limitations section where we now can support StyleSystem. Additionally we need to include documentation for using style system with examples either in the next archetype or in WKND events sample project. Additionally we need to capture the need for registering the property appliedCssClassNames against any custom sling model that needs style system capability.

[SPA Editor - Limitations](https://experienceleague.adobe.com/docs/experience-manager-64/developing/headless/spas/spa-overview.html?lang=en#limitations)

Commit:
Including appliedCssClassNames on Host Comp

Commit:
Testcase updated for better coverage

friendlymahi commented:
@niekraaijmakers - Here you go ! With my limited amt of angular knowledge i managed to make the changes in right place. Please review it from your end. However, I am not sure of the warnings or errors logged in codecov, requesting your help on this.

@CezCz

niekraaijmakers commented on src/lib/layout/aem-component.directive.ts:
```diff
@@ -198,6 +198,17 @@ export class AEMComponentDirective implements AfterViewInit, OnInit, OnDestroy,
+    const appliedCssClassNames = this.cqItem[Constants.APPLIED_CLASS_NAMES] || '';
+
+    if (appliedCssClassNames && this._component) {
+      this.renderer.setAttribute(this._component.location.nativeElement, 'class', appliedCssClassNames);
```
Can we use 'addClass' here instead of setting the attribute, overwriting it?

https://angular.io/api/core/Renderer2

If it doesn't work with addClass I am OK with it for now.

friendlymahi replied:
Already responded on this earlier !!

niekraaijmakers commented:
Left 1 comment. Rest is 100% fine and ready to go I think!

friendlymahi commented:
> Left 1 comment. Rest is 100% fine and ready to go I think!

Don't see any comment. Can you please confirm what you're referring to?

Thanks !

niekraaijmakers commented:
![Screenshot 2021-08-19 at 05 27 56](https://user-images.githubusercontent.com/26429828/130003030-434a283f-1ad1-4831-8279-636b96fa0a2d.png)

I mean here :) in the review

friendlymahi commented:
> ![Screenshot 2021-08-19 at 05 27 56](https://user-images.githubusercontent.com/26429828/130003030-434a283f-1ad1-4831-8279-636b96fa0a2d.png)
> 
> I mean here :) in the review

Thanks for the reply. I couldn't locate the view from where the review was provided. addClass doesn't help for appliedCssClassNames. Reason is, while authoring you may alter the CSS classes, and they get updated in CRX as expected, but in the markup, it shall append the 2nd class name also without wiping off the previous CSS class names. So is the reason for setting them instead of adding more. Only alternative is first set it empty or remove previous class (which may require storing the value some where) before adding new classes. Hope this clarifies. Thanks !!

niekraaijmakers commented:
Ah of course, that makes perfect sense. I also see this is the only class. Okay, let's move ahead. Let me see if I can get somebody to merge this PR.

niekraaijmakers commented:
Oh. Forgot thing.. you need to add a comitizen commit ... can you just bump the readme with a empty line.. 

and then commit with message:

feat(stylesytem): your message

?

Commit:
feat(stylesytem): Supports StyleSystem

Angular SPAs can now take advantage of StyleSystem,
and authoring experience can be enriched.

friendlymahi committed:
Merge branch 'master' into feature-stylesystem

friendlymahi commented:
> Oh. Forgot thing.. you need to add a comitizen commit ... can you just bump the readme with a empty line..
> 
> and then commit with message:
> 
> feat(stylesytem): your message
> 
> ?

Done. I guess the new commit I just did matches the pattern. Can you please confirm the same? Thanks !!

friendlymahi committed:
feat(stylesytem): Supports StyleSystem

Angular SPAs can now take advantage of StyleSystem, and authoring experience can be enriched.

friendlymahi commented:
@CezCz - Please let me know if you need any addl info for merging this PR. Thanks !!

friendlymahi commented:
@CezCz - Good morning to you !! Can you please help us push this PR merged, so we can work on the documentation to socialize this feature? Thank you in advance !!

friendlymahi commented:
@CezCz - Good Morning 2 you !! Any luck with the review? :)

friendlymahi commented:
@CezCz - Thank you so much for your support :)
